### PR TITLE
Implementing dictionary cluster labelling(issue #21)

### PR DESCRIPTION
**Pull request by Cryaaa**
Working on Issue #21, speed up through avoiding the generation of a cluster image. When in 2D rendering mode the speed up is quite noticeable, while in 3D some other rendering processes seem to dominate the computational time for larger datasets. For finalisation a method to save the actual cluster image should be added as an option.

### CODE REVIEW

**Comment by lazigu** [comment]
Hi, Ryan, great work with this! Could I ask you what do you mean by 'for finalisation a method to save the actual cluster image should be added as an option.'? Because if I understand it right, this can always be done just in napari window by clicking 'save selected layer'. Is this the same as what you meant?

**Comment by Cryaaa** [comment]
@lazigu so because the layer now is just a copy of the labels layer with a different colormap, if you save it you wont have the specific colormap saved, which visualises the clusters. It would be good to be able to save the parametric cluster image which was being generated before. Does that make more sense?

**Comment by lazigu** [comment]
Ahh yes, now I got it! It would be great to have an option to save it together with the colormap. Need to look into it.

File: napari_clusters_plotter/_kmeans_clustering.py
Comment (Czaki) [inline napari_clusters_plotter/_kmeans_clustering.py:68]: use literal dict 
```suggestion
                                                options={"min": 2, "step": 1})
```

File: napari_clusters_plotter/_kmeans_clustering.py
Comment (Czaki) [inline napari_clusters_plotter/_kmeans_clustering.py:18]:
```suggestion
DEFAULTS = {
    "kmeans_nr_clusters": 2,
    "kmeans_nr_iterations": 3000,
    "normalization": False,
    "hdbscan_min_clusters_size": 5,
}
```
Use dict syntax

File: napari_clusters_plotter/_kmeans_clustering.py
Comment (Czaki) [inline napari_clusters_plotter/_kmeans_clustering.py:56]:
```suggestion
                                                      options={"choices": ["", "KMeans", "HDBSCAN"]})
```

File: napari_clusters_plotter/_plotter.py
Comment (Czaki) [inline napari_clusters_plotter/_plotter.py:371]: You have a long list of colors but maybe there should be a cycle operation 
```suggestion
                c=[colors[int(x) % len(colors)] for x in self.cluster_ids],
```

File: napari_clusters_plotter/_plotter.py
Comment (Czaki) [inline napari_clusters_plotter/_plotter.py:398]: merge 398 and 404 lines in one clausule

File: napari_clusters_plotter/_plotter.py
Comment (Czaki) [inline napari_clusters_plotter/_plotter.py:394]: cycle list:
```suggestion
            cluster_id_dict = {i+1 : cmap[int(color) % len(cmap) ] for i,color in enumerate(np_cluster_ids_p1)}
```

**Review by Czaki — CHANGES_REQUESTED** [review]
I do a fast review with a few remarks. From my side this PR should be split into a few ones because it covers more than one topic and may make it harder to understand.

File: napari_clusters_plotter/_plotter.py
Comment (Czaki) [reply napari_clusters_plotter/_plotter.py:398]: yes. by cycling I mean protection before to high cluster number

**Comment by lazigu** [comment]
Hi @Czaki, thanks a lot for the feedback, it is very valuable for beginners like me! Actually, these changes are split into multiple PRs, but I just realized that I forgot to mention the issue related to magicgui in the PR (still getting used to Github!), so I mentioned it just now (https://github.com/BiAPoL/napari-clusters-plotter/pull/25). It just happened that this branch was branched off from the one I worked on to use magicgui, sorry for the confusion! And again, thanks a lot for the feedback!

**Comment by Czaki** [comment]
@lazigu sometimes it is better to wait until one PR is merged and then start another.

**Comment by lazigu** [comment]
> @lazigu sometimes it is better to wait until one PR is merged and then start another.

Yes, I totally see it now, it gets confusing very fast.

**Comment by lazigu** [comment]
Hi Ryan @Cryaaa, should this be marked 'ready for review'? Or are you still working on it and do you want us to have a saving cluster image option available already in this PR?

**Comment by Cryaaa** [comment]
Hey @lazigu ,
I have still have a commit open I am testing at the moment, which will implement some of the changes suggested by @Czaki . I can mark it as ready for review once I have pushed that commit :)

**Comment by Cryaaa** [comment]
@lazigu, I've implemented the changes now. I think the conflicts that are shown here will largely be resolved as soon as HDBSCAN has been merged with the main branch but I am also happy to take a look at the conflicts if some should persist :)

**Comment by Czaki** [comment]
Instead of writing your own code to convert hex color to the color tuple, you may want to use `vispy.color.Color`, which is already used by napari

For example:
```
In [4]: Color("#e377c2")
Out[4]: <Color: (0.8901961, 0.46666667, 0.7607843, 1.0)>
```

File: napari_clusters_plotter/_plotter.py
Comment (lazigu) [inline napari_clusters_plotter/_plotter.py:386]: should we leave in the code measurement of the process time or was it just for testing?

**Review by lazigu — APPROVED** [review]
I tested this and it works like a charm with no bugs from my side. I really like how noise labels are visualized as smaller dots!

File: napari_clusters_plotter/_plotter.py
Comment (Cryaaa) [reply napari_clusters_plotter/_plotter.py:386]: This was just for testing purposes, I will remove it with the next commit.

**Comment by Cryaaa** [comment]
@lazigu I implemented the change suggested by @Czaki  to use the vispy color class instead of the custom function and it everything works fine for me with these changes, so I think we are done on this branch for now!

File: napari_clusters_plotter/_plotter.py
Comment (Czaki) [reply napari_clusters_plotter/_plotter.py]: I do not know if you know about profiler and how to use it but if not then you may read a little about it here: https://napari.org/developers/profiling.html and here https://gitlab.mimuw.edu.pl/python-tools/debugger

File: napari_clusters_plotter/_plotter.py
Comment (lazigu) [inline napari_clusters_plotter/_plotter.py:397]:
```suggestion
                self.visualized_labels_layer = self.viewer.add_labels(self.analysed_layer.data, color = cluster_id_dict,
                                                                      name = "cluster_ids_in_space")
```

File: napari_clusters_plotter/_plotter.py
Comment (lazigu) [inline napari_clusters_plotter/_plotter.py:16]:
```suggestion

```
I think pandas is now unused in this file

File: napari_clusters_plotter/_plotter.py
Comment (lazigu) [inline napari_clusters_plotter/_plotter.py:13]:
```suggestion
```
and this function is not needed anymore

**Pull request merged**
